compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/Draggable.skiko.kt

```diff
@@ -0,0 +1,207 @@
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.ExperimentalComposeUiApi
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.composed
+import androidx.compose.ui.geometry.Offset
+import androidx.compose.ui.input.pointer.AwaitPointerEventScope
+import androidx.compose.ui.input.pointer.PointerButton
+import androidx.compose.ui.input.pointer.PointerEvent
+import androidx.compose.ui.input.pointer.PointerEventType
+import androidx.compose.ui.input.pointer.PointerInputChange
+import androidx.compose.ui.input.pointer.PointerKeyboardModifiers
+import androidx.compose.ui.input.pointer.PointerType
+import androidx.compose.ui.input.pointer.changedToUpIgnoreConsumed
+import androidx.compose.ui.input.pointer.pointerInput
+import androidx.compose.ui.input.pointer.positionChange
+import androidx.compose.ui.platform.LocalKeyboardModifiers
+import androidx.compose.ui.util.fastAll
+import kotlin.jvm.JvmInline
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.currentCoroutineContext
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.isActive
+
+@ExperimentalFoundationApi
+data class DragChange(
+    val offset: Offset = Offset.Zero,
+    val previousKeyboardModifiers: PointerKeyboardModifiers,
+    val currentKeyboardModifiers: PointerKeyboardModifiers
+) {
+    fun keyboardModifierChanged(): Boolean =
+        previousKeyboardModifiers != currentKeyboardModifiers
+}
+
+
+@JvmInline
+value class DragStartResult(val dragAmount: Offset = Offset.Zero)
+
+@OptIn(ExperimentalFoundationApi::class)
+typealias PointerFilter = PointerFilterScope.() -> Boolean
+
+@Suppress("MemberVisibilityCanBePrivate")
+object AwaitDragStart {
+    @OptIn(ExperimentalFoundationApi::class)
+    val Default: suspend AwaitPointerEventScope.(PointerEvent, PointerFilter) -> DragStartResult? =
+        { initialDown, filter ->
+            val awaitVariant = when (initialDown.changes[0].type) {
+                PointerType.Mouse -> OnSlop
+                else -> OnLongPress
+            }
+            awaitVariant(initialDown, filter)
+        }
+
+    @OptIn(ExperimentalFoundationApi::class)
+    val OnSlop: suspend AwaitPointerEventScope.(PointerEvent, PointerFilter) -> DragStartResult? = { initialDown, _ ->
+        var overSlop = Offset.Zero
+        var drag: PointerInputChange?
+        do {
+            drag = awaitPointerSlopOrCancellation(
+                initialDown.changes[0].id,
+                initialDown.changes[0].type
+            ) { change, over ->
+                change.consume()
+                overSlop = over
+            }
+        } while (drag != null && !drag.isConsumed)
+
+        if (drag == null) {
+            null
+        } else {
+            DragStartResult(overSlop)
+        }
+    }
+
+    @OptIn(ExperimentalComposeUiApi::class, ExperimentalFoundationApi::class)
+    val OnLongPress: suspend AwaitPointerEventScope.(PointerEvent, PointerFilter) -> DragStartResult? =
```

Comment:
It seems we also need to check if the user moves the pointer, and cancel the drag if it exceeds the slop.

Also, `draggable`, `clickable`, `toggleable` usually add new functionality (besides detecting the gesture). And usually, if we perform a drag on long click, the device informs the user with haptic feedback. I don't know if it is right to add this functionality to `draggable`, and how we can do that. What do you think?